packages/flutter_tools/lib/src/asset.dart

```diff
@@ -313,6 +314,49 @@ DevFSContent _createFontManifest(Map<String, dynamic> manifestDescriptor,
   return new DevFSStringContent(JSON.encode(fonts));
 }
 
+// Given an assets directory like this:
+//
+// assets/foo
+// assets/var1/foo
+// assets/var2/foo
+// assets/bar
+//
+// variantsFor('assets', 'foo') => ['/assets/var1/foo', '/assets/var2/foo']
+// variantsFor('assets', 'bar') => []
+class _AssetDirectoryCache {
+  _AssetDirectoryCache(Iterable<String> excluded) {
+    _excluded = excluded.map<String>((String path) => fs.path.absolute(path) + fs.path.separator);
+  }
+
+  Iterable<String> _excluded;
+  final Map<String, Map<String, List<String>>> _cache = <String, Map<String, List<String>>>{};
+
+  List<String> variantsFor(String assetPath) {
+    final String assetName = fs.path.basename(assetPath);
+    final String directory = fs.path.dirname(assetPath);
+
+    if (_cache[directory] == null) {
+      final Iterable<String> paths = fs.directory(directory).listSync(recursive: true)
+        .map((FileSystemEntity entity) => entity.path)
+        .where((String path) {
+          return fs.isFileSync(path) &&
+            !_excluded.any((String exclude) => path.startsWith(exclude));
```

Comment:
Writing a loop here saves the cost of creating a list of paths and then a copy of that after filtering out the excluded paths.  The excluded list is length 2, I think that's probably enough tweaking.